userspace/engine/rule_loader.cpp

```diff
@@ -96,67 +172,61 @@ nlohmann::json rule_loader::context::as_json()
 
 	ret["locations"] = nlohmann::json::array();
 
+	// All valid contexts should have at least one location.
 	if(m_locs.empty())
+	{
+		throw falco_exception("rule_loader::context without location?");
+	}
+
+	for(auto& loc : m_locs)
 	{
 		nlohmann::json jloc, jpos;
 
-		jloc["item_type"] = "file";
-		jloc["item_name"] = "";
+		jloc["item_type"] = item_type_as_string(loc.item_type);
+		jloc["item_name"] = loc.item_name;
 
-		jpos["filename"] = name;
-		jpos["line"] = 0;
-		jpos["column"] = 0;
-		jpos["offset"] = 0;
+		jpos["name"] = loc.name;
+		jpos["line"] = loc.pos.line;
+		jpos["column"] = loc.pos.column;
+		jpos["offset"] = loc.pos.pos;
 
 		jloc["position"] = jpos;
 
 		ret["locations"].push_back(jloc);
 	}
-	else
-	{
-		for(auto& loc : m_locs)
-		{
-			nlohmann::json jloc, jpos;
-
-			jloc["item_type"] = loc.item_type;
-			jloc["item_name"] = loc.item_name;
-
-			jpos["filename"] = name;
-			jpos["line"] = loc.mark.line;
-			jpos["column"] = loc.mark.column;
-			jpos["offset"] = loc.mark.pos;
-
-			jloc["position"] = jpos;
-
-			ret["locations"].push_back(jloc);
-		}
-	}
 
 	return ret;
 }
 
 std::string rule_loader::context::snippet(const falco::load_result::rules_contents_t& rules_contents,
 					  size_t snippet_width) const
 {
-	std::string ret;
-
+	// All valid contexts should have at least one location.
 	if(m_locs.empty())
 	{
-		return "<No context available>\n";
+		throw falco_exception("rule_loader::context without location?");
 	}
 
 	rule_loader::context::location loc = m_locs.back();
+	auto it = rules_contents.find(loc.name);
 
-	auto it = rules_contents.find(name);
+	if(alt_content.empty() && it == rules_contents.end())
+	{
+		if(it == rules_contents.end())
```

Comment:
The nested if statement seems useless here, am I wrong?

Reply:
Yeah, this is redundant. I removed it.